process_report/process_report.py

```diff
@@ -46,43 +55,53 @@ def load_institute_map() -> dict:
     return institute_map
 
 
-def load_old_pis(old_pi_file):
-    old_pi_dict = dict()
-
+def load_old_pis(old_pi_file) -> pandas.DataFrame:
     try:
-        with open(old_pi_file) as f:
-            for pi_info in f:
-                pi, first_month = pi_info.strip().split(",")
-                old_pi_dict[pi] = first_month
+        old_pi_df = pandas.read_csv(old_pi_file)
     except FileNotFoundError:
         print("Applying credit 0002 failed. Old PI file does not exist")
         sys.exit(1)
 
-    return old_pi_dict
+    return old_pi_df
 
 
-def dump_old_pis(old_pi_file, old_pi_dict: dict):
-    with open(old_pi_file, "w") as f:
-        for pi, first_month in old_pi_dict.items():
-            f.write(f"{pi},{first_month}\n")
+def dump_old_pis(old_pi_file, old_pi_df: pandas.DataFrame):
+    old_pi_df.to_csv(old_pi_file, index=False)
 
 
-def is_old_pi(old_pi_dict, pi, invoice_month):
-    first_invoice_month = old_pi_dict.get(pi, invoice_month)
-    if compare_invoice_month(first_invoice_month, invoice_month):
+def get_pi_age(old_pi_df: pandas.DataFrame, pi, invoice_month):
+    """Returns time difference between current invoice month and PI's first invoice month
+    I.e 0 for new PIs
+    Can be negative if processing past invoices"""
+    first_invoice_month = old_pi_df.loc[old_pi_df[PI_PI_FIELD] == pi, PI_FIRST_MONTH]
+    if first_invoice_month.empty:
+        return 0
+
+    month_diff = get_month_diff(invoice_month, first_invoice_month.iat[0])
+    if month_diff < 0:
         sys.exit(
             f"PI {pi} from {first_invoice_month} found in {invoice_month} invoice!"
         )
-    if compare_invoice_month(invoice_month, first_invoice_month):
-        return True
-    return False
+    else:
+        return month_diff
 
 
-def compare_invoice_month(month_1, month_2):
-    """Returns True if 1st date is later than 2nd date"""
+def get_month_diff(month_1, month_2):
+    """Returns a positive integer if month_1 is ahead in time of month_2"""
     dt1 = datetime.datetime.strptime(month_1, "%Y-%m")
     dt2 = datetime.datetime.strptime(month_2, "%Y-%m")
-    return dt1 > dt2
+    return (dt1.year - dt2.year) * 12 + (dt1.month - dt2.month)
+
+
+def add_invoice_month(invoice_month, month_to_add):
+    """Given an invoice month string, i.e "2023-02", and an integer,
+    add as many months to the invoice month, and return the sum in same
+    format"""
+    dt = datetime.datetime.strptime(invoice_month, "%Y-%m")
+    new_month = dt.month - 1 + month_to_add
+    new_year = dt.year + new_month // 12
+    new_dt = datetime.datetime(year=new_year, month=new_month % 12 + 1, day=1)
```

Comment:
it's not intuitive as to why we need to subtract 1 in line 101 and then add 1 back again.

Reply:
I didn't come up with the math for this myself, this was taken from this [StackOverflow answer](https://stackoverflow.com/a/4131114). I have added a few more testcases for this function to be sure.

Reply:
I don't understand why this is necessary? This calculation is only used for your bookkeeping and can be inferred from the age of the PI. If pi age is 0, don't read remainder, write remainder in first remainder column. if pi age is 1, read first remainder, write second remainder.

Reply:
I guess I didn't want to iterate through the old_pi dataframe to check for PIs with an age of 1. I thought writing a function to calculate and return the invoice month string would make the code more organized. Also, you said:
> If pi age is 0, don't read remainder, write remainder in first remainder column. if pi age is 1, read first remainder, write second remainder.

Just so I'm sure, you want it so that, if the pi age is 0 (new PI), and they have an entry in the PI file already (maybe because we're re-processing same invoice, different invoice for same month, etc), you want their entry to be overwritten and possibly change?

Reply:
> I guess I didn't want to iterate through the old_pi dataframe to check for PIs with an age of 1. I thought writing a function to calculate and return the invoice month string would make the code more organized. Also, you said:

Why do you need to iterate through the old_pi dataframe to check for PIs with an age of 1?

For each PI that you are processing you can find the entry in the PI file and have all the information you need. 

> 
> > If pi age is 0, don't read remainder, write remainder in first remainder column. if pi age is 1, read first remainder, write second remainder.
> 
> Just so I'm sure, you want it so that, if the pi age is 0 (new PI), and they have an entry in the PI file already (maybe because we're re-processing same invoice, different invoice for same month, etc), you want their entry to be overwritten and possibly change?

Yes, do overwrite it and print a warning if the value is different.

Reply:
> Why do you need to iterate through the old_pi dataframe to check for PIs with an age of 1?

I needed to iterate through the PI file instead of the current invoice to find these 1-month-old PIs because they might not appear (have billable projects) in the current month's invoice. If the 1-month-old PI doesn't appear in the current invoice, it's logically means they have used $0 of credit, and this is the reason why this [code block](https://github.com/CCI-MOC/process_csv_report/pull/44/files#diff-162d25340846c53be6b9c357f242c269b4564a175d9f873c6a37a63239dd5419R375-R378) (which is why `add_invoice_month()` is made) was added.

I will add warnings in the code for the warning, and now that I think about it, do you want me to using the `logging` module, instead of `print()`, or should that be a seperate PR?

Reply:
Ok, I understand the reasoning. For PIs on their second month (age == 1) and have no usage, you want to populate the value of the second remainder correctly.

I think it's a fair trade off to just leave those values Null for the sake of code simplicity. The code is already hard to follow and the value for second remainder is not used by anything.

- For each PI in current invoice
  - If age of PI == 0
    - Do calculation as before. At the end create new row in PIs.csv with current invoice month, credit and remainder of first month. Value for remainder of second month is empty.
  - If age of PI == 1
    - Read the remainder of second month from PIs.csv. Write remainder of second month. 
    - If no usage in second month, second remainder will be a null value.

Reply:
> do you want me to using the `logging` module, instead of `print()`, or should that be a seperate PR?

that can be a separate PR.

Reply:
> Ok, I understand the reasoning. For PIs on their second month (age == 1) and have no usage, you want to populate the value of the second remainder correctly.
> 
> I think it's a fair trade off to just leave those values Null for the sake of code simplicity. The code is already hard to follow and the value for second remainder is not used by anything.
> 
> * For each PI in current invoice
>   
>   * If age of PI == 0
>     
>     * Do calculation as before. At the end create new row in PIs.csv with current invoice month, credit and remainder of first month. Value for remainder of second month is empty.
>   * If age of PI == 1
>     
>     * Read the remainder of second month from PIs.csv. Write remainder of second month.
>     * If no usage in second month, second remainder will be a null value.

Alternatively, and what I proposed before and I favor, is instead of counting the `remainder` of the credit to instead use that column to track `credit used` and preinitialize both first and second month to 0 and update them as you go.

Reply:
I believe that is why we have a column to track the original value so we can calculate the remainder by subtracting the credit used from the original credit to get the credit remaining.

Reply:
@knikolla To be sure...
> Alternatively, and what I proposed before and I favor, 
Does this mean you favor your second suggestion, or first one, or both?

> If no usage in second month, second remainder will be a null value.
With the current code, whenever a new PI entry is added, their `1st Month Remainder` and `2nd Month Remainder` fields are initialized to `None` (which is empty when written out to a csv file). I take it that you don't believe I should include [this code block](https://github.com/CCI-MOC/process_csv_report/pull/44/files#diff-162d25340846c53be6b9c357f242c269b4564a175d9f873c6a37a63239dd5419R375-R378) because there's already so much going on, so I'll just remove it then. With this block removed, 1-month PIs who have no usage in the second month will have their `2nd Month Remainder` set as empty.

> preinitialize both first and second month to 0 and update them as you go.
I think if the `1st Month Remainder` and `2nd Month Remainder` fields are initialized to 0, there will be ambiguity on whether a PI spent all their credits, or if they didn't appear on their second month of invoicing?

Reply:
I was saying that's why in my initial suggestion when we were discussion the feature implementation I favor the column to mean Credit Used (1st and second month) instead of Credit Remaining (first and second month), in which case Credit Used zero would be unambiguous.

I'm okay with keeping the columns to mean Credit Remainder 1st month and Credit remainder 2nd month and for an empty value to mean credit not used.

Reply:
Could I ask why having the columns to track `Credits Used` instead of `Credits Remaining` would be better? I will follow your preference, since I don't see much difference in terms of structure or burden of maintenance for either approach.

Reply:
Mostly a gut feeling, but if I have to try to think about that preference a bit more.

- The values of credit used for the first month and the second month are independent.
- They hold meaning by themselves, as the value that we forgave from users balances for their 1st and 2nd month.
- In most cases they will match the balance for their first and (sometimes) second month unless big expenditures.
- I think of the credit used as the "active" number that is decreasing the value of the starting credit (static). So instead of writing down 1000, 400, 200 (the sequence of changing state). In my head I can better visualize it when written as 1000, -600, -200 (balance and the 2 expenditures).

Reply:
Please remove this function (and its associated test) since we're not using it anymore.